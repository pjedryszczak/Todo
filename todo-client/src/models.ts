export interface Todo {
        id?: number,
        content: string,
        todoListId: number,
        checked: boolean
    }
export interface TodoList {
        id?: number,
        userId: number,
        title: string
    }   

 