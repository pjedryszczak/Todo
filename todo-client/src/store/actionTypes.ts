export const GET_TODOLISTS = 'GET_TODOLISTS'
export const GET_TODOLISTS_SUCCESS = 'GET_TODOLISTS_SUCCESS'
export const GET_TODOLISTS_FAIL = 'GET_TODOLISTS_FAIL'
export const GET_TODOLISTS_LOADING = 'GET_TODOLISTS_LOADING'

export const SAVE_TODOLIST = 'SAVE_TODOLIST'
export const SAVE_TODOLIST_SUCCESS = 'SAVE_TODOLIST_SUCCESS'
export const SAVE_TODOLIST_FAIL = 'SAVE_TODOLIST_FAIL'
export const SAVE_TODOLIST_LOADING = 'SAVE_TODOLIST_LOADING'

export const DELETE_TODOLIST = 'DELETE_TODOLIST'
export const DELETE_TODOLIST_SUCCESS = 'DELETE_TODOLIST_SUCCESS'
export const DELETE_TODOLIST_FAIL = 'DELETE_TODOLIST_FAIL'
export const DELETE_TODOLIST_LOADING = 'DELETE_TODOLIST_LOADING'



export const GET_TODOS_FOR_LIST = 'GET_TODOS_FOR_LIST'
export const GET_TODOS_FOR_LIST_SUCCESS = 'GET_TODOS_FOR_LIST_SUCCESS'
export const GET_TODOS_FOR_LIST_FAIL = 'GET_TODOS_FOR_LIST_FAIL'
export const GET_TODOS_FOR_LIST_LOADING = 'GET_TODOS_FOR_LIST_LOADING'

export const SAVE_TODO = 'SAVE_TODO'
export const SAVE_TODO_SUCCESS = 'SAVE_TODO_SUCCESS'
export const SAVE_TODO_FAIL = 'SAVE_TODO_FAIL'
export const SAVE_TODO_LOADING = 'SAVE_TODO_LOADING'

export const DELETE_TODO = 'DELETE_TODO'
export const DELETE_TODO_SUCCESS = 'DELETE_TODO_SUCCESS'
export const DELETE_TODO_FAIL = 'DELETE_TODO_FAIL'
export const DELETE_TODO_LOADING = 'DELETE_TODO_LOADING'

export const UPDATE_TODO = 'UPDATE_TODO'
export const UPDATE_TODO_SUCCESS = 'UPDATE_TODO_SUCCESS'
export const UPDATE_TODO_FAIL = 'UPDATE_TODO_FAIL'
export const UPDATE_TODO_LOADING = 'UPDATE_TODO_LOADING'

export const CLEAR_TODOS_FOR_LIST = 'CLEAR_TODOS_FOR_LIST'

export const LOGIN_USER = 'LOGIN_USER'
export const LOGIN_USER_SUCCESS = 'LOGIN_USER_SUCCESS'
export const LOGIN_USER_FAIL = 'LOGIN_USER_FAIL'
export const LOGIN_USER_LOADING = 'LOGIN_USER_LOADING'

export const LOG_OUT = 'LOG_OUT'
export const LOG_OUT_SUCCESS = 'LOG_OUT_SUCCESS'
export const CLEAR_USER = 'CLEAR_USER'

export const REGISTER_USER = 'REGISTER_USER'
export const REGISTER_USER_SUCCESS = 'REGISTER_USER_SUCCESS'
export const REGISTER_USER_FAIL = 'REGISTER_USER_FAIL'
export const REGISTER_USER_LOADING = 'REGISTER_USER_LOADING'

export const CLEAR_ERROR = 'CLEAR_ERROR'
