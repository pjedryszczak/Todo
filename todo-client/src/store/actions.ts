import * as types from './actionTypes';
import { LoginPayload, RegisterPayload, NumericPayload, TodoListPayload, TodoPayload } from '../models';

export function getTodoListsForUser(payload: NumericPayload){
    return { type: types.GET_TODOLISTS, payload };
}
export function saveTodoList(payload: TodoListPayload){
    return { type: types.SAVE_TODOLIST, payload };
}
export function deleteTodoList(payload: TodoListPayload){
    return { type: types.DELETE_TODOLIST, payload };
}
export function getTodosForList(payload: NumericPayload){
    return { type: types.GET_TODOS_FOR_LIST, payload };
}

export function saveTodo(payload: TodoPayload){
    return { type: types.SAVE_TODO, payload };
}
export function deleteTodo(payload: TodoPayload){
    return { type: types.DELETE_TODO, payload };
}
export function updateTodo(payload: TodoPayload){
    return { type: types.UPDATE_TODO, payload };
}
export function clearTodos(){
    return { type: types.CLEAR_TODOS_FOR_LIST };
}

export function login(payload: LoginPayload){
    return { type: types.LOGIN_USER, payload };
}
export function logout(payload: any){
    return { type: types.LOG_OUT };
}
export function clearUser(){
    return { type: types.CLEAR_USER };
}
export function clearError(){
    return { type: types.CLEAR_ERROR };
}
export function register(payload: RegisterPayload){
    return { type: types.REGISTER_USER, payload };
}

