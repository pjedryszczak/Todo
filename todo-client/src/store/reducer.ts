import { TodoList, Todo } from "../models";
import * as types from './actionTypes';
export interface TodoAppState {
    todoLists: TodoList[],
    todosForSelectedTodoList: Todo[],
    loading: boolean,
    errors: Array<string>
}
export const initialState: TodoAppState = {
todoLists: [],
loading: false,
errors: [],
todosForSelectedTodoList: []
}
export function reducer(state = initialState, action: any) : TodoAppState {
    switch(action.type){
        //#region TodoList
        case types.GET_TODOLISTS_SUCCESS:
            return {
                ...state,
                todoLists: action.payload,
                loading: false,
                errors: []
            };
        case types.GET_TODOLISTS_FAIL:
            return {
                ...state,
                loading: false,
                errors: action.payload ?? []
            };
        case types.GET_TODOLISTS_LOADING:
            return {
                ...state,
                loading: true
            };            

            case types.SAVE_TODOLIST_SUCCESS:
                return {
                    ...state,
                    loading: false,
                    errors: []
                };
            case types.SAVE_TODOLIST_FAIL:
                return {
                    ...state,
                    loading: false,
                    errors: action.payload ?? []
                };
            case types.SAVE_TODOLIST_LOADING:
                return {
                    ...state,
                    loading: true
                };

            case types.DELETE_TODOLIST_SUCCESS:
                return {
                    ...state,
                    loading: false,
                    errors: []
                };
            case types.DELETE_TODOLIST_FAIL:
                return {
                    ...state,
                    loading: false,
                    errors: action.payload ?? []
                };
            case types.DELETE_TODOLIST_LOADING:
                return {
                    ...state,
                    loading: true
                };
        //#endregion
        //#region Todo
            case types.GET_TODOS_FOR_LIST_SUCCESS:
                return {
                    ...state,
                    todosForSelectedTodoList: action.payload,
                    loading: false,
                    errors: []
                };
            case types.GET_TODOS_FOR_LIST_FAIL:
                return {
                    ...state,
                    loading: false,
                    errors: action.payload ?? []
                };
            case types.GET_TODOS_FOR_LIST_LOADING:
                return {
                    ...state,
                    loading: true
                };

            case types.SAVE_TODO_SUCCESS:
                return {
                    ...state,
                    loading: false,
                    errors: []
                };
            case types.SAVE_TODO_FAIL:
                return {
                    ...state,
                    loading: false,
                    errors: action.payload ?? []
                };
            case types.SAVE_TODO_LOADING:
                return {
                    ...state,
                    loading: true
                };

            case types.DELETE_TODO_SUCCESS:
                return {
                    ...state,
                    loading: false,
                    errors: []
                };
            case types.DELETE_TODO_FAIL:
                return {
                    ...state,
                    loading: false,
                    errors: action.payload ?? []
                };
            case types.DELETE_TODO_LOADING:
                return {
                    ...state,
                    loading: true
                };

                case types.UPDATE_TODO_SUCCESS:
                    return {
                        ...state,
                        loading: false,
                        errors: []
                    };
                case types.UPDATE_TODO_FAIL:
                    return {
                        ...state,
                        loading: false,
                        errors: action.payload ?? []
                    };
                case types.UPDATE_TODO_LOADING:
                    return {
                        ...state,
                        loading: true
                    };

                case types.CLEAR_TODOS_FOR_LIST:
                    return {
                        ...state,
                        todosForSelectedTodoList: []
                    };       
            //#endregion
        default:
            return state;    
    }
}