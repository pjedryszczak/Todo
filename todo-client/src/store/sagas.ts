import { fork, put, all, takeEvery, call } from 'redux-saga/effects';
import * as types from './actionTypes';
import axios from 'axios'
import { TodoList, LoginPayload, User, RegisterPayload, NumericPayload, TodoListPayload, TodoPayload } from '../models';
import { getBaseUrl } from '../helpers';

//#region TodoList
function* getTodoLists(action: { type: string, payload: NumericPayload }){
    try {
        const token: string = localStorage.token ?? '';
        const userId = action.payload.id;
        const history = action.payload.history;
        yield put({ type: types.GET_TODOLISTS_LOADING });
        const base: string = getBaseUrl();
        const path: string = base + 'api/Todos/GetTodoLists';
        const response = yield call(() => axios.get(path, { params: {
            userId
        },
        headers: {
            'Authorization': `Bearer ${token}`,
            'X-Requested-With': 'XMLHttpRequest'
        }}));

        if(response.status === 200) {
            var results = response.data as TodoList[];
            yield put({ type: types.GET_TODOLISTS_SUCCESS, payload: results });
        }
        else if(response.status === 401){
            yield put({ type: types.LOG_OUT, history });
        }
        else{
            yield put({ type: types.GET_TODOLISTS_FAIL });
        }
    } catch (error) {
        yield put({ type: types.GET_TODOLISTS_FAIL });
    }
}

function* watchForGetTodoLists() {
    yield takeEvery(types.GET_TODOLISTS, getTodoLists);
}

function* saveTodoList(action: { type: string, payload: TodoListPayload }){
        const { todoList, history }= action.payload;
    try {
        const token: string = localStorage.token ?? '';
        yield put({ type: types.SAVE_TODOLIST_LOADING });        
        const base: string = getBaseUrl();
        const path: string = base + 'api/Todos/SaveTodoList';
        const response = yield call(() => axios.post(path, todoList,{
            headers: {
                'Authorization': `Bearer ${token}`,
                'X-Requested-With': 'XMLHttpRequest'
            }}));

        if(response.status === 200) {
            yield put({ type: types.SAVE_TODOLIST_SUCCESS });
            const payload: NumericPayload ={
                id: todoList.userId,
                history
            }
            yield put({ type: types.GET_TODOLISTS, payload: payload});
        }
        else if(response.status === 401){
            yield put({ type: types.LOG_OUT, history });
        }
        else{
            yield put({ type: types.SAVE_TODOLIST_FAIL, payload: response.data });
        }
    } catch (error) {
        let errors = error.response.data 
        yield put({ type: types.SAVE_TODOLIST_FAIL, payload: errors});
    }
}

function* watchForSaveTodoList() {
    yield takeEvery(types.SAVE_TODOLIST, saveTodoList);
}

function* deleteTodoList(action: { type: string, payload: TodoListPayload }){
    const { todoList, history }= action.payload;
    try {
        const token: string = localStorage.token ?? '';
        yield put({ type: types.DELETE_TODOLIST_LOADING });
        const todoListId = todoList.id;
        const base: string = getBaseUrl();
        const path: string = base + 'api/Todos/DeleteTodoList';
        const response = yield call(() => axios.delete(path, {params: {
            id: todoListId
        },
        headers: {
            'Authorization': `Bearer ${token}`,
            'X-Requested-With': 'XMLHttpRequest'
        }
    }));

        if(response.status === 200) {
            yield put({ type: types.DELETE_TODOLIST_SUCCESS });
            const payload: NumericPayload ={
                id: todoList.userId,
                history
            }
            yield put({ type: types.GET_TODOLISTS, payload: payload});
        }
        else if(response.status === 401){
            yield put({ type: types.LOG_OUT, history });
        }
        else{
            yield put({ type: types.DELETE_TODOLIST_FAIL, payload: response.data });
        }
    } catch (error) {
        let errors = error.response.data
        yield put({ type: types.DELETE_TODOLIST_FAIL, payload: errors});
    }
}

function* watchForDeleteTodoList() {
    yield takeEvery(types.DELETE_TODOLIST, deleteTodoList);
}

//#endregion

//#region Todo
function* getTodos(action: { type: string, payload: NumericPayload}){
    try {
        const token: string = localStorage.token ?? '';
        const todoListId = action.payload.id;
        const history = action.payload.id;
        yield put({ type: types.GET_TODOS_FOR_LIST_LOADING });
        const base: string = getBaseUrl();
        const path: string = base + 'api/Todos/GetTodos';
        const response = yield call(() => axios.get(path, { params: {
            todoListId
        },
        headers: {
            'Authorization': `Bearer ${token}`,
            'X-Requested-With': 'XMLHttpRequest'
        }
          }));

        if(response.status === 200) {
            var results = response.data as TodoList[];
            yield put({ type: types.GET_TODOS_FOR_LIST_SUCCESS, payload: results });
        }
        else if(response.status === 401){
            yield put({ type: types.LOG_OUT, history });
        }
        else{
            yield put({ type: types.GET_TODOS_FOR_LIST_FAIL, payload: response.data });
        }
    } catch (error) {
        let errors = error.response.data;
        yield put({ type: types.GET_TODOS_FOR_LIST_FAIL, payload: errors });
    }
}

function* watchForGetTodos() {
    yield takeEvery(types.GET_TODOS_FOR_LIST, getTodos);
}

function* saveTodo(action: { type: string, payload: TodoPayload }){
    try {
        const token: string = localStorage.token ?? '';
        const { todo, history }= action.payload;
        yield put({ type: types.SAVE_TODO_LOADING });        
        const base: string = getBaseUrl();
        const path: string = base + 'api/Todos/SaveTodo';
        const response = yield call(() => axios.post(path, todo, {
            headers: {
                'Authorization': `Bearer ${token}`,
                'X-Requested-With': 'XMLHttpRequest'
            }
          }));

        if(response.status === 200) {
            yield put({ type: types.SAVE_TODO_SUCCESS });
            const payload: NumericPayload = {
                id: todo.todoListId,
                history
            }
            yield put({ type: types.GET_TODOS_FOR_LIST, payload: payload });
        }
        else if(response.status === 401){
            yield put({ type: types.LOG_OUT, history });
        }
        else{
            yield put({ type: types.SAVE_TODO_FAIL, payload: response.data });
        }
    } catch (error) {
        let errors = error.response.data;
        yield put({ type: types.SAVE_TODO_FAIL, payload: errors});
    }
}

function* watchForSaveTodo() {
    yield takeEvery(types.SAVE_TODO, saveTodo);
}

function* deleteTodo(action: { type: string, payload: TodoPayload }){
    try {
        const token: string = localStorage.token ?? '';
        const { todo, history }= action.payload;
        yield put({ type: types.DELETE_TODO_LOADING });
        const base: string = getBaseUrl();
        const path: string = base + 'api/Todos/DeleteTodo';
        const response = yield call(() => axios.delete(path, { params: {
            id: todo.id ?? 0
        },
        headers: {
            'Authorization': `Bearer ${token}`,
            'X-Requested-With': 'XMLHttpRequest'
        }
    }));

        if(response.status === 200) {
            yield put({ type: types.DELETE_TODO_SUCCESS });
            const payload: NumericPayload = {
                id: todo.todoListId,
                history
            }
            yield put({ type: types.GET_TODOS_FOR_LIST, payload: payload });
        }
        else if(response.status === 401){
            yield put({ type: types.LOG_OUT, history });
        }
        else{
            yield put({ type: types.DELETE_TODO_FAIL, payload: response.data });
        }
    } catch (error) {
        let errors = error.response.data;
        yield put({ type: types.DELETE_TODO_FAIL, payload: errors});
    }
}

function* watchForDeleteTodo() {
    yield takeEvery(types.DELETE_TODO, deleteTodo);
}

function* updateTodo(action: { type: string, payload: TodoPayload }){
    try {
        const token: string = localStorage.token ?? '';
        const { todo, history }= action.payload;
        yield put({ type: types.UPDATE_TODO_LOADING });
        const base: string = getBaseUrl();
        const path: string = base + 'api/Todos/UpdateTodo';
        const response = yield call(() => axios.put(path, todo,  {
            headers: {
                'Authorization': `Bearer ${token}`,
                'X-Requested-With': 'XMLHttpRequest'
            }
          }));

        if(response.status === 200) {
            yield put({ type: types.UPDATE_TODO_SUCCESS });
            const payload: NumericPayload = {
                id: todo.todoListId,
                history
            }
            yield put({ type: types.GET_TODOS_FOR_LIST, payload: payload });
        }else if(response.status === 401){
            yield put({ type: types.LOG_OUT, history });
        }
        else{
            yield put({ type: types.UPDATE_TODO_FAIL, payload: response.data });
        }
    } catch (error) {
        let errors = error.response.data;
        yield put({ type: types.UPDATE_TODO_FAIL, payload: errors});
    }
}

function* watchForUpdateTodo() {
    yield takeEvery(types.UPDATE_TODO, updateTodo);
}
//#endregion

//#region Log
function* login(action: { type: string, payload: LoginPayload }) {
    const model = action.payload.payload;
    const history = action.payload.history;
    const base: string = getBaseUrl();
        const path: string = base + 'api/Users/Authenticate';
    
    try {
        
        const response = yield call(() => axios.post(path, model,{
            headers: {
                'X-Requested-With': 'XMLHttpRequest'
            }
        }));
    
        if(response.status === 200) {
            var user = response.data as User;

            localStorage.setItem("token", user.token);

            yield put({ type: types.LOGIN_USER_SUCCESS, payload: user });
            history.push("/home");
        }
        else if(response.status === 401){
            yield put({ type: types.LOG_OUT, history });
        }
        else{
            yield put({ type: types.LOGIN_USER_FAIL, payload: response.data });
        }
    } catch (error) {
        let errors = error.response.data;
        yield put({ type: types.LOGIN_USER_FAIL, payload: errors});
    }
}
function* watchForLogin() {
    yield takeEvery(types.LOGIN_USER, login);
}
function* register(action: { type: string, payload: RegisterPayload }) {
    const model = action.payload.payload;
    const history = action.payload.history;
    const base: string = getBaseUrl();
        const path: string = base + 'api/Users/Register';
    try {
        const response = yield call(() => axios.post(path, model, {
            headers: {
                'X-Requested-With': 'XMLHttpRequest'
            }
        }));
    
        if(response.status === 200) {
            var user = response.data as User;
            yield put({ type: types.REGISTER_USER_SUCCESS, user });
            history.push("/");
        }
        else if(response.status === 401){
            yield put({ type: types.LOG_OUT, history });
        }
        else{
            yield put({ type: types.REGISTER_USER_FAIL, payload: response.data });
        }
    } catch (error) {
        let errors = error.response.data;
        yield put({ type: types.REGISTER_USER_FAIL, payload: errors});
    }
}

function* watchForRegister() {
    yield takeEvery(types.REGISTER_USER, register);
}

function* logout(action: { type: string, payload: any }) {
    const history = action.payload;   
    yield put({ type: types.LOG_OUT_SUCCESS }); 
    history.push("/");
}

function* watchForLogout() {
    yield takeEvery(types.LOG_OUT, logout);
}
//#endregion

export default function* sagas(): any {
    yield all([
        fork(watchForGetTodoLists),
        fork(watchForSaveTodoList),
        fork(watchForDeleteTodoList),
        fork(watchForGetTodos),
        fork(watchForSaveTodo),
        fork(watchForDeleteTodo),
        fork(watchForUpdateTodo),
        fork(watchForLogin),
        fork(watchForLogout),
        fork(watchForRegister)
    ])
}