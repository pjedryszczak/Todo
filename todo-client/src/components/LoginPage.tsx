import React, { ChangeEvent, FormEvent } from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { login, logout, clearUser } from './../store/actions';
import { User, LoginModel, LoginPayload } from '../models';
import { TodoAppState } from '../store/reducer';
import { RouteComponentProps } from 'react-router';
import { sanitizeInput } from '../helpers';

interface DispatchProps {
    login: typeof login,
    logout: typeof logout,
    clearUser: typeof clearUser
}
interface StoreState {
    user: User,
    loggedIn: boolean,
    loading: boolean
}

interface LocalState {
    [key: string]: any;
    username: string,
    password: string,
    submitted: boolean,
    pushHome: boolean
}

type Props = DispatchProps & StoreState & LocalState & RouteComponentProps;
class LoginPage extends React.Component<Props, LocalState> {
    constructor(props:any) {
        super(props);

        this.props.clearUser();

        this.state = {
            username: '',
            password: '',
            submitted: false,
            pushHome: false
        };

        this.handleChange = this.handleChange.bind(this);
        this.handleSubmit = this.handleSubmit.bind(this);
    }    

    handleChange(e: ChangeEvent<HTMLInputElement>) {
        const { name, value } = e.target;
        this.setState({ [name]: value });
    }

    handleSubmit(e: FormEvent<HTMLFormElement>) {
        e.preventDefault();

        this.setState({ submitted: true });
        const { username, password } = this.state;
        const { history } = this.props;
        if (username && password) {
            const loginPayload: LoginModel ={
                username,
                password,
            }
            const sagaPayload: LoginPayload = {
                payload: loginPayload,                
                history
            }
            this.props.login(sagaPayload);
        }
    }
    handleKeyDown(e: any) {
        sanitizeInput(e);
    }

    render() {
        const { loading } = this.props;
        const { username, password, submitted } = this.state;
        return (
            <div>
                <h2>Login</h2>
                <form className="col 12" name="form" onSubmit={this.handleSubmit} onKeyDown={this.handleKeyDown}>
                    <div className={'form-group' + (submitted && !username ? ' has-error' : '')}>
                        <label htmlFor="username">Username</label>
                        <input type="text" className="form-control" name="username" value={username} onChange={this.handleChange} />
                        {submitted && !username &&
                        <span data-error="wrong" >Usarname is required</span>
                        }
                    </div>
                    <div className={'form-group' + (submitted && !password ? ' has-error' : '')}>
                        <label htmlFor="password">Password</label>
                        <input type="password" className="form-control" name="password" value={password} onChange={this.handleChange} />
                        {submitted && !password &&
                            <span data-error="wrong" >Password is required</span>
                        }
                    </div>
                    <div className="form-group">
                        <button className="waves-effect waves-light blue darken-1 btn">Login</button>
                        {loading &&
                            
                           <>≈ÅADOWANIE</>
                        }
                        <Link to="/register" className="waves-effect waves-light green darken-1 btn">Register</Link>
                    </div>
                </form>
            </div>
        );
    }
}
function mapStateToProps(state: TodoAppState, props: any){
    return {
        loading: state.loading
    }
}
const container = connect(
    mapStateToProps,
    {
    login,
    logout,
    clearUser
}
)(LoginPage)

export default container;