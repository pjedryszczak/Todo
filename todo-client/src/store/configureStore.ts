import {createStore, applyMiddleware } from "redux";
import { composeWithDevTools } from "redux-devtools-extension";
import thunk from "redux-thunk";
import createSagaMiddleware from "redux-saga";
import sagas from "./sagas";
import { reducer } from "./reducer";

export default function configureStore(initialState: any, history: any) {
    const composeEnhancers = composeWithDevTools({});
    const reduxSagaMiddleware = createSagaMiddleware();

    const middleware = [reduxSagaMiddleware, thunk ];
    const applyMiddlewareFunc = applyMiddleware(...middleware);

    let store: any;
    if(process.env.NODE_ENV === 'production') {
        store = createStore(reducer, applyMiddlewareFunc);
    }else{
        store = createStore(reducer,initialState, composeEnhancers(applyMiddlewareFunc));
    }

    reduxSagaMiddleware.run(sagas);

    return store;
}