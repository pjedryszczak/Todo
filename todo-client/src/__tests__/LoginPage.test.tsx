import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import configureStore from 'redux-mock-store';
import renderer from 'react-test-renderer';
import { clearUser } from './../store/actions'

import LoginPage from '../components/LoginPage';
import { Provider } from 'react-redux';
import { User } from '../models';
import { StaticRouter } from 'react-router-dom';

const mockStore = configureStore([]);
describe('LoginPage', () => {
    let store: any;
    let component: renderer.ReactTestRenderer; 
   const user: User ={
    firstName: 'TestName',
    id: 0,
    username: 'testname',
    token: ''
   }
    beforeEach(() => {
      store = mockStore({
        loggedIn: false,
        user: user,
        loading: false
      });
      store.dispatch = jest.fn();

      component = renderer.create(
          <StaticRouter>
        <Provider store={store}>
            <LoginPage />
        </Provider>
        </StaticRouter>);
    });

    it('it renders with given state from Redux store', () => {
        expect(component.toJSON()).toMatchSnapshot();
    });
    it('clear action on mount', () => {
             
        expect(store.dispatch).toHaveBeenCalledTimes(1);
        expect(store.dispatch).toHaveBeenCalledWith(
            clearUser()
        );
      });
  });