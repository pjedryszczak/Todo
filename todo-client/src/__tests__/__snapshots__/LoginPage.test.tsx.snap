// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage it renders with given state from Redux store 1`] = `
<div>
  <h2>
    Login
  </h2>
  <form
    className="col 12"
    name="form"
    onKeyDown={[Function]}
    onSubmit={[Function]}
  >
    <div
      className="form-group"
    >
      <label
        htmlFor="username"
      >
        Username
      </label>
      <input
        className="form-control"
        name="username"
        onChange={[Function]}
        type="text"
        value=""
      />
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="password"
      >
        Password
      </label>
      <input
        className="form-control"
        name="password"
        onChange={[Function]}
        type="password"
        value=""
      />
    </div>
    <div
      className="form-group"
    >
      <button
        className="waves-effect waves-light blue darken-1 btn"
      >
        Login
      </button>
      <a
        className="waves-effect waves-light green darken-1 btn"
        href="/register"
        onClick={[Function]}
      >
        Register
      </a>
    </div>
  </form>
</div>
`;
